clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2245,6 +2245,36 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
 
     break;
   }
+  case Builtin::BI__builtin_hlsl_and: {
+    if (SemaRef.checkArgCount(TheCall, 2))
+      return true;
+    if (CheckVectorElementCallArgs(&SemaRef, TheCall))
+      return true;
+
+    // CheckVectorElementCallArgs(...) guarantees both args are the same type.
+    assert(TheCall->getArg(0)->getType() == TheCall->getArg(1)->getType() &&
+           "Both args must be of the same type");
+
+    // check that the arguments are bools or, if vectors,
+    // vectors of bools
+    QualType ArgTy = TheCall->getArg(0)->getType();
+    if (const auto *VecTy = ArgTy->getAs<VectorType>()) {
+      ArgTy = VecTy->getElementType();
+    }
+    if (!getASTContext().hasSameUnqualifiedType(ArgTy,
+                                                getASTContext().BoolTy)) {
+      SemaRef.Diag(TheCall->getBeginLoc(),
+                   diag::err_typecheck_convert_incompatible)
```

Comment:
@Icohedron, I think we need to figure out a better diagnostic for this to emit. Perhaps using `err_typecheck_expect_scalar_or_vector`. Using that diagnostic you can simplify this code by calling `CheckScalarOrVector` to verify that the arguments are bool or bool vector.